frontend/src/metabase-lib/v1/expressions/pratt/compiler.ts

```diff
@@ -83,35 +107,75 @@ function compileValue(
   };
 }
 
-function compileField(node: Node): Lib.ExpressionParts {
-  assert(node.type === FIELD, "Invalid node type");
-  assert(node.token?.text, "Empty field name");
-  assert(node.token?.value, "Empty field value");
+function getKindForType(type: ExpressionType): Kind {
+  switch (type) {
+    case "boolean":
+      return "segment";
+    case "aggregation":
+      return "metric";
+    default:
+      return "field";
+  }
+}
 
-  return withNode(node, {
-    operator: "dimension" as Lib.ExpressionOperator,
-    options: {},
-    args: [node.token.value],
-  });
+function compileDimension(name: string, node: Node, ctx: Context) {
+  assert(typeof name === "string", t`Invalid dimension name: ${name}`);
+
+  if (!ctx.resolver) {
+    return {
+      operator: "dimension" as Lib.ExpressionOperator,
+      options: {},
+      args: [name],
+    };
+  }
+
+  try {
+    const kind = getKindForType(ctx.type);
+    const dimension = ctx.resolver(kind, name, node);
+    return withNode(node, dimension);
+  } catch (err) {
+    const operator = getMBQLName(name);
```

Comment:
We can remove this branch once https://github.com/metabase/metabase/pull/56835 lands.